src/impls.rs

```diff
@@ -113,200 +90,119 @@ impl_metadata_for_tuple!(A, B, C, D, E, F, G, H);
+	fn type_info() -> Type {
+		TypeVariant::new("Result", Namespace::prelude())
+			.type_params(tuple_meta_type!(T, E))
+			.variants(
+				Variants::with_fields()
+					.variant("Ok", Fields::unnamed().field_of::<T>())
+					.variant("Err", Fields::unnamed().field_of::<E>()),
+			)
+			.into()
 	}
 }
 
-impl<K, V> HasTypeDef for BTreeMap<K, V>
+impl<K, V> TypeInfo for BTreeMap<K, V>
 where
 	K: Metadata + 'static,
 	V: Metadata + 'static,
 {
-	fn type_def() -> TypeDef {
-		TypeDefStruct::new(vec![NamedField::new("elems", MetaType::new::<[(K, V)]>())]).into()
-	}
-}
-
-impl<T> HasTypeId for Box<T>
-where
-	T: HasTypeId + ?Sized,
-{
-	fn type_id() -> TypeId {
-		T::type_id()
+	fn type_info() -> Type {
+		TypeComposite::new("BTreeMap", Namespace::prelude())
+			.type_params(tuple_meta_type![(K, V)])
+			.fields(Fields::named().field_of::<[(K, V)]>("elems"))
+			.into()
 	}
 }
 
-impl<T> HasTypeDef for Box<T>
+impl<T> TypeInfo for Box<T>
 where
-	T: Metadata + ?Sized,
+	T: TypeInfo + ?Sized,
 {
-	fn type_def() -> TypeDef {
-		T::type_def()
+	fn type_info() -> Type {
+		T::type_info()
 	}
 }
 
-impl<T> HasTypeId for &T
+impl<T> TypeInfo for &T
 where
-	T: HasTypeId + ?Sized,
+	T: TypeInfo + ?Sized,
 {
-	fn type_id() -> TypeId {
-		T::type_id()
+	fn type_info() -> Type {
+		T::type_info()
 	}
 }
 
-impl<T> HasTypeDef for &T
+impl<T> TypeInfo for &mut T
 where
-	T: Metadata + ?Sized,
+	T: TypeInfo + ?Sized,
 {
-	fn type_def() -> TypeDef {
-		T::type_def()
+	fn type_info() -> Type {
+		T::type_info()
 	}
 }
 
-impl<T> HasTypeId for &mut T
-where
-	T: HasTypeId + ?Sized,
-{
-	fn type_id() -> TypeId {
-		T::type_id()
-	}
-}
-
-impl<T> HasTypeDef for &mut T
-where
-	T: Metadata + ?Sized,
-{
-	fn type_def() -> TypeDef {
-		T::type_def()
-	}
-}
-
-impl<T> HasTypeId for [T]
+impl<T> TypeInfo for [T]
 where
 	T: Metadata + 'static,
 {
-	fn type_id() -> TypeId {
-		TypeIdSlice::of::<T>().into()
-	}
-}
-
-impl<T> HasTypeDef for [T]
-where
-	T: Metadata,
-{
-	fn type_def() -> TypeDef {
-		TypeDef::builtin()
-	}
-}
-
-impl HasTypeId for str {
-	fn type_id() -> TypeId {
-		TypeIdPrimitive::Str.into()
+	fn type_info() -> Type {
+		TypeSlice::of::<T>().into()
 	}
 }
 
-impl HasTypeDef for str {
-	fn type_def() -> TypeDef {
-		TypeDef::builtin()
+impl TypeInfo for str {
+	fn type_info() -> Type {
+		TypePrimitive::Str.into()
 	}
 }
 
-impl HasTypeId for String {
-	fn type_id() -> TypeId {
-		<str>::type_id()
-	}
-}
-
-impl HasTypeDef for String {
-	fn type_def() -> TypeDef {
-		TypeDefStruct::new(vec![NamedField::new("vec", MetaType::new::<Vec<u8>>())]).into()
-	}
-}
-
-impl<T> HasTypeId for PhantomData<T>
-where
-	T: Metadata + ?Sized,
-{
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("PhantomData", Namespace::prelude(), vec![T::meta_type()]).into()
+impl TypeInfo for String {
+	fn type_info() -> Type {
+		TypeComposite::new("String", Namespace::prelude())
+			.fields(Fields::named().field_of::<Vec<u8>>("vec"))
+			.into()
```

Comment:
Either we go full way and represent `String` the same as `str` or we do the same as I proposed for `HashMap` where we make it like a tuple struct to be able to drop the `vec` name for the field.

Reply:
Maybe we actually can do better than this? What we actually want is like a type alias, or what do you think? So we could say that a `String` is really just a `str` but still something different.

Reply:
> Either we go full way and represent `String` the same as `str` or we do the same as I proposed for `HashMap` where we make it like a tuple struct to be able to drop the `vec` name for the field.

Note that the difference here was that the actual representation of `String` is a struct with `vec: Vec<u8>`, which is not true for `HashMap`. 

However I made this change *before* we decided on making this lib to be specific to describing `SCALE`.

> Maybe we actually can do better than this? What we actually want is like a type alias, or what do you think? So we could say that a String is really just a str but still something different.

A good idea because this way we are not losing information i.e. that the original type was `String` and not `str`.

Reply:
> A good idea because this way we are not losing information i.e. that the original type was String and not str.

Yes, but actually I am asking myself if this is useful information with `scale-info` being a SCALE codec protocol, not a type protocol.

Reply:
https://github.com/paritytech/scale-info/issues/9